File: apps/web/src/api/mirror.ts
import { rocksky } from "../lib/rocksky";

export type MirrorProvider = "lastfm" | "listenbrainz" | "tealfm";

export interface MirrorSourceView {
  provider: MirrorProvider;
  enabled: boolean;
  externalUsername?: string;
  hasCredentials: boolean;
  lastPolledAt?: string;
  lastScrobbleSeenAt?: string;
}

export const getMirrorSources = async (): Promise<MirrorSourceView[]> => {
  const res = await rocksky().mirrorSources();
  return (res.sources ?? []) as MirrorSourceView[];
};

export interface PutMirrorSourceInput {
  provider: MirrorProvider;
  enabled?: boolean;
  externalUsername?: string;
  /** Omit to leave existing key unchanged. Empty string to clear. */
  apiKey?: string;
}

export const putMirrorSource = async (
  input: PutMirrorSourceInput,
): Promise<MirrorSourceView> => {
  const res = await rocksky().putMirrorSource(input);
  return res as MirrorSourceView;
};

File: sdk/swift/Sources/Rocksky/Models/Mirror.swift
import Foundation

public struct MirrorSourceView: Codable, Hashable, Sendable {
    public let provider: String
    public let enabled: Bool
    public let externalUsername: String?
    public let hasCredentials: Bool
    public let lastPolledAt: String?
    public let lastScrobbleSeenAt: String?
}

public struct MirrorSourcesResponse: Codable, Sendable {
    public let sources: [MirrorSourceView]
}

public struct PutMirrorSourceInput: Codable, Sendable {
    public let provider: String
    public let enabled: Bool?
    public let externalUsername: String?
    /// API key / token. Omit to leave unchanged; empty string to clear.
    public let apiKey: String?

    public init(
        provider: String,
        enabled: Bool? = nil,
        externalUsername: String? = nil,
        apiKey: String? = nil
    ) {
        self.provider = provider
        self.enabled = enabled
        self.externalUsername = externalUsername
        self.apiKey = apiKey
    }
}
